qsimcirq/qsim_simulator.py

```diff
@@ -62,6 +64,120 @@ def __init__(self, qsim_options: dict = {}):
+        repetitions: The number of samples to take.
+
+    Returns:
+        A dictionary from measurement gate key to measurement
+        results. Measurement results are stored in a 2-dimensional
+        numpy array, the first dimension corresponding to the repetition
+        and the second to the actual boolean measurement results (ordered
+        by the qubits being measured.)
+
+    Raises:
+        NotImplementedError: If there are non-terminal measurements in the
+            circuit.
+        ValueError: If there are multiple MeasurementGates with the same key,
+            or if repetitions is negative.
+    """
+    if not isinstance(program, qsimc.QSimCircuit):
+      program = qsimc.QSimCircuit(program, device=program.device)
+
+    if program.are_all_measurements_terminal() and repetitions > 1:
+      print('Provided circuit has no intermediate measurements. ' +
+            'It may be faster to sample from the final state vector. ' +
+            'Continuing with one-by-one sampling.')
+
+    # Compute indices of measured qubits
+    ordered_qubits = ops.QubitOrder.DEFAULT.order_for(program.all_qubits())
+    ordered_qubits = list(reversed(ordered_qubits))
+
+    qubit_map = {
+      qubit: index for index, qubit in enumerate(ordered_qubits)
+    }
+
+    # Computes
+    # - the list of qubits to be measured
+    # - the start (inclusive) and end (exclusive) indices of each measurement
+    # - a mapping from measurement key to measurement gate
+    measurement_ops = [
+      op for _, op, _ in program.findall_operations_with_gate_type(ops.MeasurementGate)
+    ]
+    measured_qubits = []  # type: List[ops.Qid]
+    bounds = {}  # type: Dict[str, Tuple]
+    meas_ops = {}  # type: Dict[str, cirq.MeasurementGate]
+    current_index = 0
+    for op in measurement_ops:
+      gate = op.gate
+      key = protocols.measurement_key(gate)
+      meas_ops[key] = gate
+      if key in bounds:
+        raise ValueError("Duplicate MeasurementGate with key {}".format(key))
+      bounds[key] = (current_index, current_index + len(op.qubits))
+      measured_qubits.extend(op.qubits)
+      current_index += len(op.qubits)
+
+    indices = [qubit_map[qubit] for qubit in measured_qubits]
+
+    # Want to check all bitstrings
+    n_qubits = len(program.all_qubits())
+    bitstrings = [i for i in range(2**n_qubits)]
+    bitstrings = [format(bs, 'b').zfill(n_qubits) for bs in bitstrings]
+
+    # Set qsim options
+    options = {}
+    options.update(self.qsim_options)
+    options['c'] = program.translate_cirq_to_qsim(ops.QubitOrder.DEFAULT)
+
+    results = {}
+    for key, bound in bounds.items():
+      boundlen = bound[1]-bound[0]
+      results[key] = np.ndarray(shape=(repetitions, bound[1]-bound[0]),
+                                dtype=int)
+
+    for i in range(repetitions):
+      measurements = qsim.qsim_sample(options)
+      print(f'measurements: {measurements}')
+      for key, bound in bounds.items():
```

Comment:
If the keys are iterated in the order of occurrence in the circuit then I guess all this might work correctly without any modifications here and in `pybind_main.cpp`.

Reply:
I'm not sure we can rely on this being the case. Cirq is pretty permissive with how users can insert measurements into the circuit, so I'd feel more comfortable keeping this conversion explicit.

Reply:
Disregard my comment. It should be OK even if the keys are not ordered.